File: python/modules/admin/api_role.py
from flask import request

from service.base import Base
from service.admin_token import AdminToken
from model.api_role import ApiRole as ApiRoleM
from model.api_menu import ApiMenu
from util.util import Util

class ApiRole(Base):

  __menus = {}    #全部菜单
  __permAll = {}  #用户权限

  # 列表
  def List(self):
    # 参数
    json = self.Json()
    token = self.JsonName(json, 'token')
    data = self.JsonName(json, 'data')
    page = self.JsonName(json, 'page')
    limit = self.JsonName(json, 'limit')
    order = self.JsonName(json, 'order')
    # 验证
    msg = AdminToken.Verify(token, request.path)
    if msg != '' : return self.GetJSON({'code':4001, 'msg':msg})
    if not data or not page or not limit :
      return self.GetJSON({'code':4000, 'msg':'参数错误!'})
    # 条件
    param = Util.JsonDecode(data)
    where, whereData = self.__getWhere(param)
    # 统计
    m = ApiRoleM()
    m.Columns('count(*) AS num')
    m.Where(where, whereData)
    total = m.FindFirst()
    # 查询
    m.Columns('id', 'name', 'FROM_UNIXTIME(ctime, %s) as ctime', 'FROM_UNIXTIME(utime, %s) as utime', 'perm')
    m.Where(where, '%Y-%m-%d %H:%i:%s', '%Y-%m-%d %H:%i:%s', whereData)
    m.Order(order if order else 'id DESC')
    m.Page(int(page), int(limit))
    list = m.Find()
    # 返回
    return self.GetJSON({'code':0, 'msg':'成功', 'list':list, 'total':int(total['num'])})

  # 搜索条件
  def __getWhere(self, param):
    # 参数
    name = Util.Trim(param['name']) if 'name' in param.keys() else ''
    # 条件
    where = 'name LIKE %s'
    whereData = ('%'+name+'%')
    return where, whereData

  # 添加
  def Add(self):
    # 参数
    json = self.Json()
    token = self.JsonName(json, 'token')
    data = self.JsonName(json, 'data')
    # 验证
    msg = AdminToken.Verify(token, request.path)
    if msg != '' :
      return self.GetJSON({'code':4001, 'msg':msg})
    if not data :
      return self.GetJSON({'code':4000, 'msg':'参数错误!'})
    # 数据
    param = Util.JsonDecode(data)
    name = Util.Trim(param['name']) if 'name' in param.keys() else ''
    if name == '' :
      return self.GetJSON({'code':4000, 'msg':'名称不能为空!'})
    # 模型
    m = ApiRoleM()
    m.Values({'name': name, 'ctime': Util.Time()})
    if m.Insert() :
      return self.GetJSON({'code':0, 'msg':'成功'})
    else :
      return self.GetJSON({'code':5000, 'msg':'添加失败!'})

  # 编辑
  def Edit(self):
    # 参数
    json = self.Json()
    token = self.JsonName(json, 'token')
    id = self.JsonName(json, 'id')
    data = self.JsonName(json, 'data')
    # 验证
    msg = AdminToken.Verify(token, request.path)
    if msg != '' :
      return self.GetJSON({'code':4001, 'msg':msg})
    if not id or not data :
      return self.GetJSON({'code':4000, 'msg':'参数错误!'})
    # 数据
    param = Util.JsonDecode(data)
    name = Util.Trim(param['name']) if 'name' in param.keys() else ''
    if name == '' :
      return self.GetJSON({'code':4000, 'msg':'名称不能为空!'})
    # 模型
    m = ApiRoleM()
    m.Set({'name': name, 'utime': Util.Time()})
    m.Where('id=%s', id)
    if m.Update() :
      return self.GetJSON({'code':0, 'msg':'成功'})
    else :
      return self.GetJSON({'code':5000, 'msg':'更新失败!'})

  # 删除
  def Del(self):
    # 参数
    json = self.Json()
    token = self.JsonName(json, 'token')
    data = self.JsonName(json, 'data')
    # 验证
    msg = AdminToken.Verify(token, request.path)
    if msg != '' :
      return self.GetJSON({'code':4001, 'msg':msg})
    if not data :
      return self.GetJSON({'code':4000, 'msg':'参数错误!'})
    # 数据
    param = Util.JsonDecode(data)
    ids = Util.Implode(',', param)
    # 模型
    m = ApiRoleM()
    m.Where('id in('+ids+')')
    if m.Delete() :
      return self.GetJSON({'code':0, 'msg':'成功'})
    else :
      return self.GetJSON({'code':5000, 'msg':'删除失败!'})

  # 权限
  def Perm(self):
    # 参数
    json = self.Json()
    token = self.JsonName(json, 'token')
    id = self.JsonName(json, 'id')
    perm = self.JsonName(json, 'perm')
    # 验证
    msg = AdminToken.Verify(token, request.path)
    if msg != '' :
      return self.GetJSON({'code':4001, 'msg':msg})
    if not id :
      return self.GetJSON({'code':4000, 'msg':'参数错误!'})
    # 模型
    m = ApiRoleM()
    m.Set({'perm': perm, 'utime': Util.Time()})
    m.Where('id=%s', id)
    if m.Update() :
      return self.GetJSON({'code':0, 'msg':'成功'})
    else :
      return self.GetJSON({'code':5000, 'msg':'更新失败!'})

  # 权限-列表
  def RoleList(self):
    # 参数
    json = self.Json()
    token = self.JsonName(json, 'token')
    # 验证
    msg = AdminToken.Verify(token, '')
    if msg != '' :
      return self.GetJSON({'code':4001, 'msg':msg})
    # 查询
    m = ApiRoleM()
    m.Columns('id', 'name')
    data = m.Find()
    lists = [{'label':'无', 'value':0}]
    for val in data :
      lists += [{'label': val['name'], 'value':val['id']}]
    return self.GetJSON({'code':0, 'msg':'成功', 'list':lists})

  # 权限-列表
  def PermList(self):
    # 参数
    json = self.Json()
    token = self.JsonName(json, 'token')
    perm = self.JsonName(json, 'perm')
    # 验证
    msg = AdminToken.Verify(token, '')
    if msg != '' :
      return self.GetJSON({'code':4001, 'msg':msg})
    # 全部菜单
    self.__menus = {}
    model = ApiMenu()
    model.Columns('id', 'fid', 'title', 'url', 'ico', 'controller', 'action')
    model.Order('sort, id')
    data = model.Find()
    for val in data :
      fid = str(val['fid'])
      if fid in self.__menus : self.__menus[fid] += [val]
      else : self.__menus[fid] = [val]
    # 用户权限
    self.__permAll = self.__permArr(perm)
    # 返回
    return self.GetJSON({'code':0, 'msg':'成功', 'list':self._getMenu('0')})

  # 权限-拆分
  def __permArr(self, perm: str):
    permAll = {}
    arr = [] if not perm else Util.Explode(' ', perm)
    for val in arr :
      s = Util.Explode(':', val)
      permAll[s[0]] = int(s[1])
    return permAll
  
  # 递归菜单
  def _getMenu(self, fid: str):
    data = []
    M = self.__menus[fid] if fid in self.__menus else []
    for val in M :
      # 菜单权限
      id = str(val['id'])
      perm = self.__permAll[id] if id in self.__permAll.keys() else 0
      # 动作权限
      action = []
      actionArr = []
      actionStr = str(val['action'])
      if actionStr != '' : actionArr = Util.JsonDecode(actionStr)
      for v in actionArr :
        permVal = int(v['perm'])
        checked = True if perm&permVal>0 else False
        action += [{
          'id': str(val['id'])+'_'+str(v['perm']),
          'label': str(v['name']),
          'checked': checked,
          'perm': v['perm'],
        }]
      # 数据
      checked = True if id in self.__permAll.keys() else False
      tmp = {'id': val['id'], 'label': val['title'], 'checked': checked}
      if val['fid']==0 : tmp['show'] = True
      # children
      menu = self._getMenu(id)
      if len(menu)>0 : tmp['children'] = menu
      elif len(action)>0 :
        tmp['action'] = True
        tmp['children'] = action
      data += [tmp]
    return data

File: java/src/main/java/webmis/modules/admin/ApiRole.java
package webmis.modules.admin;

import java.util.ArrayList;
import java.util.HashMap;

import javax.servlet.http.HttpServletRequest;

import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;

import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import webmis.model.ApiMenu;
import webmis.service.AdminToken;
import webmis.service.Base;
import webmis.util.Type;
import webmis.util.Util;

@RestController
@Controller("AdminApiRole")
@RequestMapping("/admin/api_role")
public class ApiRole extends Base {

  private static String where;
  private static Object[] whereData;
  private static HashMap<String, ArrayList<HashMap<String, Object>>> menus = null;  //全部菜单
  private static HashMap<String, Long> permAll = null;                              //用户权限

  /* 列表 */
  @RequestMapping("list")
  String List(@RequestBody JSONObject json, HttpServletRequest request) {
    HashMap<String,Object> res;
    // 参数
    String token = JsonName(json, "token");
    String data = JsonName(json, "data");
    int page = Integer.valueOf(JsonName(json, "page"));
    int limit = Integer.valueOf(JsonName(json, "limit"));
    String order = JsonName(json, "order");
    // 验证
    String msg = AdminToken.Verify(token, request.getRequestURI());
    if(!msg.equals("")){
      res = new HashMap<String,Object>();
      res.put("code", 4001);
      res.put("msg", msg);
      return GetJSON(res);
    }
    if(data=="" || page==0 || limit==0){
      res = new HashMap<String,Object>();
      res.put("code", 4000);
      res.put("msg", "参数错误!");
      return GetJSON(res);
    }
    // 条件
    JSONObject param = Util.JsonDecode(data);
    whereData = getWhere(param);
    // 统计
    webmis.model.ApiRole m = new webmis.model.ApiRole();
    m.Columns("count(*) AS num");
    m.Where(where, whereData);
    HashMap<String, Object> total = m.FindFirst();
    // 查询
    m.Columns("id", "name", "FROM_UNIXTIME(ctime, '%Y-%m-%d %H:%i:%s') as ctime", "FROM_UNIXTIME(utime, '%Y-%m-%d %H:%i:%s') as utime", "perm");
    m.Where(where, whereData);
    m.Order(!order.equals("")?order:"id DESC");
    m.Page(page, limit);
    ArrayList<HashMap<String,Object>> list = m.Find();
    // 返回
    res = new HashMap<String,Object>();
    res.put("code", 0);
    res.put("msg", "成功");
    res.put("list", list);
    res.put("total", Type.Int(total.get("num")));
    return GetJSON(res);
  }
  /* 搜索条件 */
  private Object[] getWhere(JSONObject param) {
    // 参数
    String name = param.containsKey("name")?String.valueOf(param.get("name")).trim():"";
    // 条件
    where = "name LIKE ?";
    Object[] whereData = {"%"+name+"%"};
    return whereData;
  }

  /* 添加 */
  @RequestMapping("add")
  String Add(@RequestBody JSONObject json, HttpServletRequest request) {
    HashMap<String,Object> res;
    // 参数
    String token = JsonName(json, "token");
    String data = JsonName(json, "data");
    // 验证
    String msg = AdminToken.Verify(token, request.getRequestURI());
    if(!msg.equals("")){
      res = new HashMap<String,Object>();
      res.put("code", 4001);
      res.put("msg", msg);
      return GetJSON(res);
    }
    if(data==""){
      res = new HashMap<String,Object>();
      res.put("code", 4000);
      res.put("msg", "参数错误!");
      return GetJSON(res);
    }
    // 数据
    JSONObject param = Util.JsonDecode(data);
    String name = param.containsKey("name")?String.valueOf(param.get("name")).trim():"";
    if(name.equals("")){
      res = new HashMap<String,Object>();
      res.put("code", 4000);
      res.put("msg", "名称不能为空!");
      return GetJSON(res);
    }
    // 模型
    webmis.model.ApiRole m = new webmis.model.ApiRole();
    HashMap<String,Object> uData = new HashMap<String,Object>();
    uData.put("name", name);
    uData.put("ctime", Util.Time());
    m.Values(uData);
    if(m.Insert()){
      res = new HashMap<String,Object>();
      res.put("code", 0);
      res.put("msg", "成功");
    } else {
      res = new HashMap<String,Object>();
      res.put("code", 5000);
      res.put("msg", "添加失败!");
    }
    return GetJSON(res);
  }

  /* 编辑 */
  @RequestMapping("edit")
  String Edit(@RequestBody JSONObject json, HttpServletRequest request) {
    HashMap<String,Object> res;
    // 参数
    String token = JsonName(json, "token");
    String id = JsonName(json, "id");
    String data = JsonName(json, "data");
    // 验证
    String msg = AdminToken.Verify(token, request.getRequestURI());
    if(!msg.equals("")){
      res = new HashMap<String,Object>();
      res.put("code", 4001);
      res.put("msg", msg);
      return GetJSON(res);
    }
    if(id=="" || data==""){
      res = new HashMap<String,Object>();
      res.put("code", 4000);
      res.put("msg", "参数错误!");
      return GetJSON(res);
    }
    // 数据
    JSONObject param = Util.JsonDecode(data);
    String name = param.containsKey("name")?String.valueOf(param.get("name")).trim():"";
    if(name.equals("")){
      res = new HashMap<String,Object>();
      res.put("code", 4000);
      res.put("msg", "名称不能为空!");
      return GetJSON(res);
    }
    // 模型
    webmis.model.ApiRole m = new webmis.model.ApiRole();
    HashMap<String,Object> uData = new HashMap<String,Object>();
    uData.put("name", name);
    uData.put("utime", Util.Time());
    m.Set(uData);
    m.Where("id=?", id);
    if(m.Update()){
      res = new HashMap<String,Object>();
      res.put("code", 0);
      res.put("msg", "成功");
    } else {
      res = new HashMap<String,Object>();
      res.put("code", 5000);
      res.put("msg", "更新失败!");
    }
    return GetJSON(res);
  }

  /* 删除 */
  @RequestMapping("del")
  String Del(@RequestBody JSONObject json, HttpServletRequest request) {
    HashMap<String,Object> res;
    // 参数
    String token = JsonName(json, "token");
    String data = JsonName(json, "data");
    // 验证
    String msg = AdminToken.Verify(token, request.getRequestURI());
    if(!msg.equals("")){
      res = new HashMap<String,Object>();
      res.put("code", 4001);
      res.put("msg", msg);
      return GetJSON(res);
    }
    if(data==""){
      res = new HashMap<String,Object>();
      res.put("code", 4000);
      res.put("msg", "参数错误!");
      return GetJSON(res);
    }
    // 数据
    JSONArray param = Util.JsonDecodeArray(data);
    String ids = Util.Implode(",", JSONArray.parseArray(param.toJSONString()));
    // 模型
    webmis.model.ApiRole m = new webmis.model.ApiRole();
    m.Where("id in("+ids+")");
    if(m.Delete()){
      res = new HashMap<String,Object>();
      res.put("code", 0);
      res.put("msg", "成功");
    } else {
      res = new HashMap<String,Object>();
      res.put("code", 5000);
      res.put("msg", "删除失败!");
    }
    return GetJSON(res);
  }

  /* 权限 */
  @RequestMapping("perm")
  String Perm(@RequestBody JSONObject json, HttpServletRequest request) {
    HashMap<String,Object> res;
    // 参数
    String token = JsonName(json, "token");
    String id = JsonName(json, "id");
    String perm = JsonName(json, "perm");
    // 验证
    String msg = AdminToken.Verify(token, request.getRequestURI());
    if(!msg.equals("")){
      res = new HashMap<String,Object>();
      res.put("code", 4001);
      res.put("msg", msg);
      return GetJSON(res);
    }
    if(id==null){
      res = new HashMap<String,Object>();
      res.put("code", 4000);
      res.put("msg", "参数错误!");
      return GetJSON(res);
    }
    // 模型
    webmis.model.ApiRole m = new webmis.model.ApiRole();
    HashMap<String,Object> uData = new HashMap<String,Object>();
    uData.put("perm", perm);
    uData.put("utime", Util.Time());
    m.Set(uData);
    m.Where("id=?", id);
    if(m.Update()){
      res = new HashMap<String,Object>();
      res.put("code", 0);
      res.put("msg", "成功");
    } else {
      res = new HashMap<String,Object>();
      res.put("code", 5000);
      res.put("msg", "更新失败!");
    }
    return GetJSON(res);
  }

  /* 角色-列表 */
  @RequestMapping("roleList")
  String RoleList(@RequestBody JSONObject json, HttpServletRequest request) {
    HashMap<String,Object> res;
    // 参数
    String token = JsonName(json, "token");
    // 验证
    String msg = AdminToken.Verify(token, "");
    if(!msg.equals("")){
      res = new HashMap<String,Object>();
      res.put("code", 4001);
      res.put("msg", msg);
      return GetJSON(res);
    }
    // 查询
    webmis.model.ApiRole m = new webmis.model.ApiRole();
    m.Columns("id", "name");
    HashMap<String, Object> tmp = new HashMap<String, Object>();
    ArrayList<HashMap<String, Object>> data = m.Find();
    ArrayList<HashMap<String, Object>> lists = new ArrayList<HashMap<String, Object>>();
    tmp.put("label", "无");
    tmp.put("value", 0);
    lists.add(tmp);
    for (HashMap<String, Object> val : data) {
      tmp = new HashMap<String, Object>();
      tmp.put("label", val.get("name"));
      tmp.put("value", val.get("id"));
      lists.add(tmp);
    }
    res = new HashMap<String,Object>();
    res.put("code", 0);
    res.put("msg", "成功");
    res.put("list", lists);
    return GetJSON(res);
  }

  /* 权限-列表 */
  @RequestMapping("permList")
  String PermList(@RequestBody JSONObject json, HttpServletRequest request) {
    HashMap<String,Object> res;
    // 参数
    String token = JsonName(json, "token");
    String perm = JsonName(json, "perm");
    // 验证
    String msg = AdminToken.Verify(token, "");
    if(!msg.equals("")){
      res = new HashMap<String,Object>();
      res.put("code", 4001);
      res.put("msg", msg);
      return GetJSON(res);
    }
    // 全部菜单
    ArrayList<HashMap<String, Object>> tmp;
    menus = new HashMap<String, ArrayList<HashMap<String, Object>>>();
    ApiMenu model = new ApiMenu();
    model.Columns("id", "fid", "title", "url", "ico", "controller", "action");
    model.Order("sort, id");
    ArrayList<HashMap<String, Object>> all = model.Find();
    for (HashMap<String, Object> val : all) {
      String fid = String.valueOf(val.get("fid"));
      if (menus.containsKey(fid)) {
        menus.get(fid).add(val);
      } else {
        tmp = new ArrayList<HashMap<String, Object>>();
        tmp.add(val);
        menus.put(fid,tmp);
      }
    }
    // 用户权限
    permAll = permArr(perm);
    // 返回
    res = new HashMap<String,Object>();
    res.put("code", 0);
    res.put("msg", "成功");
    res.put("list", _getMenu("0"));
    return GetJSON(res);
  }
  // 权限-拆分
  private HashMap<String, Long> permArr(String perm) {
    HashMap<String, Long> permAll = new HashMap<String, Long>();
    ArrayList<String> arr = !perm.isEmpty()?Util.Explode(" ", perm):new ArrayList<String>();
    ArrayList<String> s;
    for(String val : arr){
      s = Util.Explode(":", val);
      permAll.put(s.get(0), Long.valueOf(s.get(1)));
    }
    return permAll;
  }
  // 递归菜单
  private ArrayList<HashMap<String, Object>> _getMenu(String fid) {
    HashMap<String, Object> tmp;
    ArrayList<HashMap<String, Object>> menu;
    ArrayList<HashMap<String, Object>> data = new ArrayList<HashMap<String, Object>>();
    ArrayList<HashMap<String, Object>> M = menus.containsKey(fid)?menus.get(fid):data;
    for( HashMap<String, Object> val : M) {
      // 菜单权限
      String id = val.get("id").toString();
      Long perm = permAll.containsKey(id)?permAll.get(id):0;
      // 动作权限
      HashMap<String, Object> actionTmp;
      ArrayList<HashMap<String, Object>> action = new ArrayList<HashMap<String, Object>>();
      String actionStr = val.get("action").toString();
      JSONArray actionArr = new JSONArray();
      if(!actionStr.equals("")) actionArr = Util.JsonDecodeArray(actionStr);
      for(int i=0; i<actionArr.size(); i++){
        Long permVal = Long.valueOf(actionArr.getJSONObject(i).get("perm").toString());
        boolean checked = (perm&permVal)>0?true:false;
        actionTmp = new HashMap<String, Object>();
        actionTmp.put("id", String.valueOf(val.get("id").toString())+"_"+String.valueOf(actionArr.getJSONObject(i).get("perm").toString()));
        actionTmp.put("label", actionArr.getJSONObject(i).get("name"));
        actionTmp.put("checked", checked);
        actionTmp.put("perm", actionArr.getJSONObject(i).get("perm"));
        action.add(actionTmp);
      }
      // 数据
      boolean checked = permAll.containsKey(id)?true:false;
      tmp = new HashMap<String, Object>();
      tmp.put("id", val.get("id"));
      tmp.put("label", val.get("title"));
      tmp.put("checked", checked);
      if(val.get("fid").equals(0)) tmp.put("show", true);
      // children
      menu = _getMenu(id);
      if(menu.size()>0) tmp.put("children", menu);
      else if(action.size()>0) {
        tmp.put("action", true);
        tmp.put("children", action);
      }
      data.add(tmp);
    }
    return data;
  }
  
}
